Revert "ntdll: Use process_vm_writev() for other process write" to restore FreeBSD support.

--- dlls/ntdll/unix/virtual.c	2025-08-15 13:42:59.000000000 -0700
+++ dlls/ntdll/unix/virtual.c	2025-08-14 13:08:13.000000000 -0700
@@ -6971,8 +6971,7 @@ NTSTATUS WINAPI NtWriteVirtualMemory( HANDLE process, 
             req->handle     = wine_server_obj_handle( process );
             req->addr       = wine_server_client_ptr( addr );
             wine_server_add_data( req, buffer, size );
-            status = wine_server_call( req );
-            size = reply->written;
+            if ((status = wine_server_call( req ))) size = 0;
         }
         SERVER_END_REQ;
     }
--- server/process.c	2025-08-15 13:42:59.000000000 -0700
+++ server/process.c	2025-08-14 13:08:13.000000000 -0700
@@ -1163,8 +1163,8 @@ int set_process_debug_flag( struct process *process, i
         peb32 = process->peb + 0x1000;
 
     /* BeingDebugged flag is the byte at offset 2 in the PEB */
-    if (peb32 && !write_process_memory( process, peb32 + 2, 1, &data, NULL )) return 0;
-    return write_process_memory( process, process->peb + 2, 1, &data, NULL );
+    if (peb32 && !write_process_memory( process, peb32 + 2, 1, &data )) return 0;
+    return write_process_memory( process, process->peb + 2, 1, &data );
 }
 
 /* create a new process */
@@ -1712,7 +1712,7 @@ DECL_HANDLER(write_process_memory)
     if ((process = get_process_from_handle( req->handle, PROCESS_VM_WRITE )))
     {
         data_size_t len = get_req_data_size();
-        if (len) write_process_memory( process, req->addr, len, get_req_data(), &reply->written );
+        if (len) write_process_memory( process, req->addr, len, get_req_data() );
         release_object( process );
     }
 }
--- server/process.h	2025-08-15 13:42:59.000000000 -0700
+++ server/process.h	2025-08-14 13:08:13.000000000 -0700
@@ -137,8 +137,7 @@ extern int read_process_memory( struct process *proces
 extern void init_process_tracing( struct process *process );
 extern void finish_process_tracing( struct process *process );
 extern int read_process_memory( struct process *process, client_ptr_t ptr, data_size_t size, char *dest );
-extern int write_process_memory( struct process *process, client_ptr_t ptr, data_size_t size, const char *src,
-                                 data_size_t *written );
+extern int write_process_memory( struct process *process, client_ptr_t ptr, data_size_t size, const char *src );
 
 static inline process_id_t get_process_id( struct process *process ) { return process->id; }
 
--- server/protocol.def	2025-08-15 13:42:59.000000000 -0700
+++ server/protocol.def	2025-08-14 13:08:13.000000000 -0700
@@ -1979,8 +1979,6 @@ struct process_info
     obj_handle_t handle;       /* process handle */
     client_ptr_t addr;         /* addr to write to */
     VARARG(data,bytes);        /* data to write */
-@REPLY
-    data_size_t written;       /* number of bytes written. */
 @END
 
 
--- server/ptrace.c	2025-08-15 13:42:59.000000000 -0700
+++ server/ptrace.c	2025-08-14 13:08:13.000000000 -0700
@@ -45,8 +45,6 @@
 #endif
 #include <unistd.h>
 
-#include <sys/uio.h>
-
 #include "ntstatus.h"
 #define WIN32_NO_STATUS
 #include "winternl.h"
@@ -334,6 +332,22 @@ static int read_thread_int( struct thread *thread, voi
     return ret;
 }
 
+/* write a long to a thread address space */
+static long write_thread_long( struct thread *thread, void *addr, unsigned long data, unsigned long mask )
+{
+    unsigned long old_data;
+    int res;
+
+    if (mask != ~0ul)
+    {
+        if (read_thread_long( thread, addr, &old_data ) == -1) return -1;
+        data = (data & mask) | (old_data & ~mask);
+    }
+    if ((res = ptrace( PTRACE_POKEDATA, get_ptrace_pid(thread), (caddr_t)addr, data )) == -1)
+        file_set_error();
+    return res;
+}
+
 /* return a thread of the process suitable for ptracing */
 static struct thread *get_ptrace_thread( struct process *process )
 {
@@ -414,13 +428,31 @@ int read_process_memory( struct process *process, clie
     return !len;
 }
 
+/* make sure we can write to the whole address range */
+/* len is the total size (in longs) */
+static int check_process_write_access( struct thread *thread, long *addr, data_size_t len )
+{
+    int page = get_page_size() / sizeof(long);
+
+    for (;;)
+    {
+        if (write_thread_long( thread, addr, 0, 0 ) == -1) return 0;
+        if (len <= page) break;
+        addr += page;
+        len -= page;
+    }
+    return (write_thread_long( thread, addr + len - 1, 0, 0 ) != -1);
+}
+
 /* write data to a process memory space */
-int write_process_memory( struct process *process, client_ptr_t ptr, data_size_t size, const char *src,
-                          data_size_t *written )
+int write_process_memory( struct process *process, client_ptr_t ptr, data_size_t size, const char *src )
 {
     struct thread *thread = get_ptrace_thread( process );
-    struct iovec local, remote;
-    ssize_t len;
+    int ret = 0;
+    long data = 0;
+    data_size_t len;
+    long *addr;
+    unsigned long first_mask, first_offset, last_mask, last_offset;
 
     if (!thread) return 0;
 
@@ -430,22 +462,74 @@ int write_process_memory( struct process *process, cli
         return 0;
     }
 
-    if (thread->unix_pid == -1 || !is_process_init_done(thread->process))
+    /* compute the mask for the first long */
+    first_mask = ~0;
+    first_offset = ptr % sizeof(long);
+    memset( &first_mask, 0, first_offset );
+
+    /* compute the mask for the last long */
+    last_offset = (size + first_offset) % sizeof(long);
+    if (!last_offset) last_offset = sizeof(long);
+    last_mask = 0;
+    memset( &last_mask, 0xff, last_offset );
+
+    addr = (long *)(unsigned long)(ptr - first_offset);
+    len = (size + first_offset + sizeof(long) - 1) / sizeof(long);
+
+    if (suspend_for_ptrace( thread ))
     {
-        set_error( STATUS_ACCESS_DENIED );
-        return 0;
+        if (!check_process_write_access( thread, addr, len ))
+        {
+            set_error( STATUS_ACCESS_DENIED );
+            goto done;
+        }
+
+        if (len > 3)
+        {
+            char procmem[24];
+            int fd;
+
+            snprintf( procmem, sizeof(procmem), "/proc/%u/mem", process->unix_pid );
+            if ((fd = open( procmem, O_WRONLY )) != -1)
+            {
+                ssize_t r = pwrite( fd, src, size, ptr );
+                close( fd );
+                if (r == size)
+                {
+                    ret = 1;
+                    goto done;
+                }
+            }
+        }
+
+        /* first word is special */
+        if (len > 1)
+        {
+            memcpy( (char *)&data + first_offset, src, sizeof(long) - first_offset );
+            src += sizeof(long) - first_offset;
+            if (write_thread_long( thread, addr++, data, first_mask ) == -1) goto done;
+            first_offset = 0;
+            len--;
+        }
+        else last_mask &= first_mask;
+
+        while (len > 1)
+        {
+            memcpy( &data, src, sizeof(long) );
+            src += sizeof(long);
+            if (write_thread_long( thread, addr++, data, ~0ul ) == -1) goto done;
+            len--;
+        }
+
+        /* last word is special too */
+        memcpy( (char *)&data + first_offset, src, last_offset - first_offset );
+        if (write_thread_long( thread, addr, data, last_mask ) == -1) goto done;
+        ret = 1;
+
+    done:
+        resume_after_ptrace( thread );
     }
-    local.iov_len = remote.iov_len = size;
-    local.iov_base = (void *)src;
-    remote.iov_base = (void *)(unsigned long)ptr;
-    len = process_vm_writev( thread->unix_pid, &local, 1, &remote, 1, 0);
-    if (written) *written = len == -1 ? 0 : len;
-    if (len == -1 || len != size)
-    {
-        set_error( STATUS_PARTIAL_COPY );
-        return 0;
-    }
-    return 1;
+    return ret;
 }
 
 /* retrieve an LDT selector entry */
